core/src/SYCL/Kokkos_SYCL_Instance.hpp

```diff
@@ -70,18 +70,157 @@ class SYCLInternal {
 
   std::unique_ptr<cl::sycl::queue> m_queue;
 
+  // USMObjectMem is a reusable buffer for a single object
+  // in USM memory
+  template <sycl::usm::alloc Kind>
+  class USMObjectMem {
+   public:
+    static constexpr sycl::usm::alloc kind = Kind;
+
+    friend void swap(USMObjectMem& lhs, USMObjectMem& rhs) noexcept {
+      using std::swap;
+      swap(lhs.m_q, rhs.m_q);
+      swap(lhs.m_data, rhs.m_data);
+      swap(lhs.m_capacity, rhs.m_capacity);
+    }
+
+    USMObjectMem()                    = default;
+    USMObjectMem(USMObjectMem const&) = delete;
+    USMObjectMem& operator=(USMObjectMem const&) = delete;
+
+    USMObjectMem(USMObjectMem&& that) noexcept
+        : m_q(std::move(that.m_q)),
+          m_data(that.m_data),
+          m_capacity(that.m_capacity) {
+      that.m_data     = nullptr;
+      that.m_capacity = 0;
+    }
+
+    USMObjectMem& operator=(USMObjectMem&& that) noexcept {
+      swap(*this, that);
+      return *this;
+    }
```

Comment:
No.  When the right hand side goes out of scope, it would `sycl::free` the allocation (which the left hand side now refers to) in its destructor.  The left hand side would also have its previous allocation leaked.